src/ir_pattern.py

```diff
@@ -0,0 +1,446 @@
+                continue
+            states = state.permute_and_match_inputrefs(
+                tail_inputs,
+                flow_graph.instr_inputs[cand_ref],
+            )
+            try_matches.extend(states)
+
+        # Continue matching by working backwards through the pattern
+        for pat_ref in body_refs[::-1]:
+            pat_inputs = pattern.flow_graph.instr_inputs[pat_ref]
+
+            next_try_matches = []
+            for state in try_matches:
+                # By pattern construction, pat_ref should be in the state's ref_map
+                # This would be true for "disjoint" or irrelevant instructions in the
+                # pattern, like random nops.
+                cand = state.try_map_ref(pat_ref)
+                if not isinstance(cand, InstrRef):
+                    continue
+                if not state.match_instr(pat_ref.instruction, cand.instruction):
+                    continue
+                states = state.permute_and_match_inputrefs(
+                    pat_inputs,
+                    flow_graph.instr_inputs[cand],
+                )
+                next_try_matches.extend(states)
+            try_matches = next_try_matches
+
+        for n, state in enumerate(try_matches):
+            # Perform any additional pattern-specific validation or computation
+            if not pattern.check(state):
+                continue
+
+            # Determine which instructions we can replace with the replacement_instr or nops
+            refs_to_replace: List[InstrRef] = [state.map_ref(tail_ref)]
+            for pat_ref in body_refs[::-1]:
+                cand_ref = state.map_ref(pat_ref)
+                # The candidate instruction can be replaced if all of its downstream
+                # uses are also being replaced
+                if all(
+                    all(r in refs_to_replace for r in refs)
+                    for _, refs in flow_graph.instr_uses[cand_ref].items()
+                ):
+                    refs_to_replace.append(cand_ref)
+
+            # Create temporary registers for the inputs to the replacement_instr
+            for pat_ref in input_refs:
+                assert len(pat_ref.instruction.outputs) == 1
+                input_reg = pat_ref.instruction.outputs[0]
+                assert isinstance(input_reg, Register)
+
+                original_reg = state.map_reg(input_reg)
+                temp_reg = Register(
+                    f"{original_reg.register_name}_fictive_{replace_index}"
+                )
+                state.rename_reg(input_reg, temp_reg)
+                move_instr = arch.parse(
+                    "move.fictive", [temp_reg, original_reg], InstructionMeta.missing()
+                )
+                input_uses = pattern.flow_graph.instr_uses[pat_ref].get(input_reg)
+                assert len(input_uses) >= 1
+                for use in input_uses:
+                    state.map_ref(use).add_instruction_before(move_instr)
```

Comment:
Needs to update `instr_uses`/`instr_inputs` as well

Reply:
I don't think this works. With the following pattern:
```python
    # inputs: $a, $b; output: $x
    replacement = "foo $x, $a, $b"
    parts = [
        "addis $p, $a, $b",
        "xoris $q, $a, $b",
        "subf $x, $p, $q",
    ]
```

This could match instructions in either `addis; xoris; subf` or `xoris; addis; subf` order in the original asm. And this move needs to come before whichever is first in the control flow.

I don't think `input_uses[0]` is guaranteed to come "first" -- at least, it's not an intended guarantee of `RefSet`.

I can try to make this more clear with a comment

Reply:
No, I think it should work: the new fictive register is only read by the final instruction. A comment is probably a good idea in any case.

Reply:
You're right; I refactored it to only insert one move instruction per temp.